modules/qna/src/backend/transfer.ts

```diff
@@ -1,66 +1,58 @@
 import _ from 'lodash'
 
-import * as parsers from './parsers.js'
+import { QnaItem, QnaEntry } from './qna'
+import Storage from './storage'
 
-const ANSWERS_SPLIT_CHAR = '†'
-
-export const importQuestions = async (questions, params) => {
-  const { storage, config, format = 'json', statusCallback, uploadStatusId } = params
+function assertIsArrayOfQnaItems(data: QnaItem[]): data is QnaItem[] {
+  if (!_.isArray(data)) {
+    return false
+  }
+  return data.reduce((accumulator: boolean, current: QnaItem) => accumulator && assertIsQnaItem(current), true)
+}
 
-  statusCallback(uploadStatusId, 'Calculating diff with existing questions')
+function assertIsQnaItem(data: QnaItem): data is QnaItem {
+  return _.isString(data.id) && assertIsQnaEntry(data.data)
+}
 
-  const existingQuestions = (await storage.fetchAllQuestions()).map(item =>
-    JSON.stringify(_.omit(item.data, 'enabled'))
+function assertIsQnaEntry(data: QnaEntry): data is QnaEntry {
+  return (
+    data &&
+    (data.action === 'text' || data.action === 'redirect' || data.action === 'text_redirect') &&
+    data.answers !== undefined &&
+    _.isString(data.category) &&
+    data.questions !== undefined &&
+    _.isString(data.redirectFlow) &&
+    _.isString(data.redirectNode)
   )
```

Comment:
We already use Joi to validate if the specified data matches our schema. It can replace all "assert*" methods. it's in qna.d.ts. 

I'm a bit puzzled with the "data is QnaEntry" part.. shouldn't it be "as" ?

Reply:
Yeah sure, I'll use Joi.

The function signature I used is what Typescript call a User Defined Type Guard. See here:

https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

It looks like this:

function isFish(pet: Fish | Bird): pet is Fish {
    return (<Fish>pet).swim !== undefined;
}

The 'instanceof' operator only works for classes and instances of classes. User defined type guards are usefull when using plain javascript objects and interfaces.